src/apps/vpn/platforms/linux/daemon/apptracker.cpp

```diff
@@ -189,15 +189,25 @@ void AppTracker::cgroupsChanged(const QString& directory) {
       Q_ASSERT(m_runningApps.contains(scope));
       AppData* data = m_runningApps.take(scope);
 
-      emit appTerminated(data->cgroup, data->appId);
+      emit appTerminated(data->cgroup(), data->appId());
       delete data;
     }
   }
 }
 
+QString AppData::getDesktopFileId(const QString& filePath) {
+  QStringList parts =
+      filePath.chopped(filePath.endsWith(".desktop") ? 8 : 0).split("/");
+  // id is "-" separated path relative to applications directory, w/o extension
+  // desktop files can also be copied in autostart directory
+  int from = std::max(parts.lastIndexOf("applications"),
+                      parts.lastIndexOf("autostart"));
```

Comment:
Could it be that form is -1?

Reply:
Yes from could be `-1`, which works as expected with `.sliced(0)`. Should I document that?

Reply:
Note that correct paths always contain a `applications` or `autostart` directory though. Allowing `-1` basically means the function is idempotent and we don’t behave unexpectedly on incorrect paths.

Reply:
I clarified to move the `from + 1` to the definition of from. Hopefully that makes it easier to read.